### ISSUE TITLE
Out of bounds error

### ISSUE THREAD

**Issue by zsunberg** [JuliaPOMDP/MCTS.jl#63]
A student in the DMU class encountered this error:

```
┌ Warning: Wall clock time limit of 0.05 seconds for choosing an action online exceeded. Taking a random action.
└ @ DMUStudent.HW3 none:14
Progress: 93%|███████████████████████████████████████ | ETA: 0:00:10
BoundsError: attempt to access 420-element Array{Float64,1} at index [290150432]
Stacktrace:
[1] getindex at .\array.jl:744 [inlined]
[2] q at C:\Users\hulse\.julia\packages\MCTS\eX6AJ\src\vanilla.jl:151 [inlined]
[3] best_sanode_UCB(::StateNode{StaticArrays.SArray{Tuple{2},Int64,1,2},Symbol}, ::Float64) at C:\Users\hulse\.julia\packages\MCTS\eX
6AJ\src\vanilla.jl:402
[4] simulate(::MCTSPlanner{DenseGridWorld,StaticArrays.SArray{Tuple{2},Int64,1,2},Symbol,MCTS.SolvedRolloutEstimator{RandomPolicy{Ran
dom._GLOBAL_RNG,DenseGridWorld,BeliefUpdaters.NothingUpdater},Random._GLOBAL_RNG},Random._GLOBAL_RNG}, ::StateNode{StaticArrays.SArray
{Tuple{2},Int64,1,2},Symbol}, ::Int64) at C:\Users\hulse\.julia\packages\MCTS\eX6AJ\src\vanilla.jl:290
[5] simulate(::MCTSPlanner{DenseGridWorld,StaticArrays.SArray{Tuple{2},Int64,1,2},Symbol,MCTS.SolvedRolloutEstimator{RandomPolicy{Ran
dom._GLOBAL_RNG,DenseGridWorld,BeliefUpdaters.NothingUpdater},Random._GLOBAL_RNG},Random._GLOBAL_RNG}, ::StateNode{StaticArrays.SArray
{Tuple{2},Int64,1,2},Symbol}, ::Int64) at C:\Users\hulse\.julia\packages\MCTS\eX6AJ\src\vanilla.jl:302 (repeats 19 times)
[6] build_tree(::MCTSPlanner{DenseGridWorld,StaticArrays.SArray{Tuple{2},Int64,1,2},Symbol,MCTS.SolvedRolloutEstimator{RandomPolicy{R
andom._GLOBAL_RNG,DenseGridWorld,BeliefUpdaters.NothingUpdater},Random._GLOBAL_RNG},Random._GLOBAL_RNG}, ::StaticArrays.SArray{Tuple{2
},Int64,1,2}) at C:\Users\hulse\.julia\packages\MCTS\eX6AJ\src\vanilla.jl:267
[7] plan! at C:\Users\hulse\.julia\packages\MCTS\eX6AJ\src\vanilla.jl:242 [inlined]
[8] action_info(::MCTSPlanner{DenseGridWorld,StaticArrays.SArray{Tuple{2},Int64,1,2},Symbol,MCTS.SolvedRolloutEstimator{RandomPolicy{
Random._GLOBAL_RNG,DenseGridWorld,BeliefUpdaters.NothingUpdater},Random._GLOBAL_RNG},Random._GLOBAL_RNG}, ::StaticArrays.SArray{Tuple{
2},Int64,1,2}) at C:\Users\hulse\.julia\packages\MCTS\eX6AJ\src\vanilla.jl:197
[9] action at C:\Users\hulse\.julia\packages\MCTS\eX6AJ\src\vanilla.jl:202 [inlined]
[10] action(::DMUStudent.HW3.HW3Policy{MCTSPlanner{DenseGridWorld,StaticArrays.SArray{Tuple{2},Int64,1,2},Symbol,MCTS.SolvedRolloutEs
timator{RandomPolicy{Random._GLOBAL_RNG,DenseGridWorld,BeliefUpdaters.NothingUpdater},Random._GLOBAL_RNG},Random._GLOBAL_RNG},NTuple{4
,Symbol}}, ::StaticArrays.SArray{Tuple{2},Int64,1,2}) at .\none:11
[11] action_info at C:\Users\hulse\.julia\packages\POMDPModelTools\R58YX\src\info.jl:12 [inlined]
[12] iterate(::POMDPSimulators.MDPSimIterator{(:s, :a),DenseGridWorld,DMUStudent.HW3.HW3Policy{MCTSPlanner{DenseGridWorld,StaticArray
s.SArray{Tuple{2},Int64,1,2},Symbol,MCTS.SolvedRolloutEstimator{RandomPolicy{Random._GLOBAL_RNG,DenseGridWorld,BeliefUpdaters.NothingU
pdater},Random._GLOBAL_RNG},Random._GLOBAL_RNG},NTuple{4,Symbol}},MersenneTwister,StaticArrays.SArray{Tuple{2},Int64,1,2}}, ::Tuple{In
t64,StaticArrays.SArray{Tuple{2},Int64,1,2}}) at C:\Users\hulse\.julia\packages\POMDPSimulators\E5RFy\src\stepthrough.jl:51
[13] evaluate(::MCTSSolver, ::String) at .\none:54
[14] #submit#1(::String, ::typeof(submit), ::MCTSSolver, ::String, ::String) at C:\Users\hulse\.julia\packages\DMUStudent\sJ097\src\D
MUStudent.jl:83
[15] (::DMUStudent.var"#kw##submit")(::NamedTuple{(:nickname,),Tuple{String}}, ::typeof(submit), ::MCTSSolver, ::String, ::String) at
.\none:0
[16] top-level scope at In[20]:1
```

**Comment by Shushman** [comment]
That's weird. The number seems to suggest an undefined location. But the `n.id` that `[290150432]` represents ultimately comes from `child_ids`. And `child_ids` is only ever populated by `length(tree.n)`
https://github.com/JuliaPOMDP/MCTS.jl/blob/master/src/vanilla.jl#L339

Sounds like a job for the debugger

**Comment by zhousy80** [comment]
When I use MCTS to solve a problem with large state space, I often meet this error. I wonder if there is any progress in addressing this issue.

**Comment by BoZenKhaa** [comment]
Can you provide a working example of the instance you get the error on?

**Comment by zhousy80** [comment]
Thanks for the help. Below is a relatively simple example:

using POMDPs, QuickPOMDPs, MCTS, DiscreteValueIteration, POMDPTools
using Distributions, Combinatorics, StaticArrays, POMDPSimulators
using TickTock

NumberUnits = 5;
Number_level = 4;
Transition_matrix = [0.86 0.14 0.0 0.0; 0.0 0.83 0.17 0.0;0.0 0.0 0.89 0.11;
1.0 0.0 0.0 0.0];
failure_penalty = -800.0;
maintenance_penalty = -200.0;
setup_cost = -200.0;
normal_operation = 50.0;

function get_state_index(sp)
    n = sum([(sp[k]-1)*Number_level^(k-1) for k=1:NumberUnits])+1
    return n
end

function get_state_vec(s)
    temp = digits(s-1, base=Number_level, pad=16)
    local_s = temp[1:NumberUnits].+1
    return local_s
end

function get_action_vec(a)
    temp = digits(a-1, base=2, pad=16)  #convert index to binary bits
    local_a = temp[1:NumberUnits];
    return local_a
end

function set_action_value(a)
    n = sum([(a[k])*2^(k-1) for k=1:NumberUnits])+1
    return n
end


multiunit1 = QuickMDP(
    gen = function (s, a, rng)
        #convert an action index to a unit number vector
        local_a = get_action_vec(a);
        local_s = get_state_vec(s);
        sp = repeat(1:1,NumberUnits);
        has_repair = false;
        r = 0.0;
        for i in 1:NumberUnits # for each units
            if local_a[i] == 0       # do nothing
                crd = Categorical(Transition_matrix[local_s[i],:]);
                sp[i] = rand(crd);
                if local_s[i]==Number_level
                    r = r + failure_penalty;
                    if has_repair==false
                        r = r+setup_cost;
                        has_repair = true;
                    end
                else
                    r = r + normal_operation;
                end
            end

            if local_a[i]==1
                sp[i] == 1; # bring it to new condition
                if local_s[i]==Number_level
                    r = r+failure_penalty;
                    if has_repair==false
                        r = r+setup_cost;
                        has_repair = true;
                    end
                else
                    r = r+maintenance_penalty;
                    if has_repair==false
                        r = r+setup_cost;
                        has_repair = true;
                    end
                end
            end
        end
        #convert sp to state index
        #println(sp)
        n = get_state_index(sp)
        return (sp=n, r=r)
    end,
    actions = 1:(2^NumberUnits), #largest 63 units
    states = 1:(Number_level^NumberUnits),
    initialstate = function()
            ImplicitDistribution() do rng
            return (1)
        end
    end, #all units start fresh
    discount = 0.95,
    isterminal = false              # no ending
)



repetition = 128;
simsteps = 20;
rewards = zeros(repetition,1)
for i in 1:repetition
println("Start MCTS simulation...", i)
mcts = MCTSSolver(n_iterations=10, depth=100, exploration_constant=10000.0);
policy = MCTSPlanner(mcts, multiunit1);
hr = HistoryRecorder(max_steps=simsteps);
history = simulate(hr, multiunit1, policy);
rewards[i,1] = discounted_reward(history);
end

mean(rewards)


After runing in Julia 1.7, I got the following error message:

Start MCTS simulation...1
ERROR: BoundsError: attempt to access 11-element Vector{Vector{Int64}} at
index [257]
Stacktrace:
  [1] getindex
    @ .\array.jl:861 [inlined]
  [2] child_ids
    @ C:\Users\szhou\.julia\packages\MCTS\vamO5\src\vanilla.jl:146 [inlined]
  [3] children
    @ C:\Users\szhou\.julia\packages\MCTS\vamO5\src\vanilla.jl:147 [inlined]
  [4] best_sanode_Q(snode::StateNode{Int64, Int64})
    @ MCTS C:\Users\szhou\.julia\packages\MCTS\vamO5\src\vanilla.jl:384
  [5]
action_info(p::MCTSPlanner{QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"),
Int64, Int64, NamedTuple{(:stateindex, :isterminal, :actionindex,
:initialstate, :gen, :actions, :states, :discount), Tuple{Dict{Int64,
Int64}, Bool, Dict{Int64, Int64}, var"#30#33", var"#29#32",
UnitRange{Int64}, UnitRange{Int64}, Float64}}}, Int64, Int64,
MCTS.SolvedRolloutEstimator{RandomPolicy{Random._GLOBAL_RNG,
QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}}, NothingUpdater}, Random._GLOBAL_RNG}, Random._GLOBAL_RNG},
s::Int64)
    @ MCTS C:\Users\szhou\.julia\packages\MCTS\vamO5\src\vanilla.jl:204
  [6] iterate(it::POMDPTools.Simulators.MDPSimIterator{(:s, :a, :sp, :r,
:info, :t, :action_info),
QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}},
MCTSPlanner{QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64,
Int64, NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate,
:gen, :actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool,
Dict{Int64, Int64}, var"#30#33", var"#29#32", UnitRange{Int64},
UnitRange{Int64}, Float64}}}, Int64, Int64,
MCTS.SolvedRolloutEstimator{RandomPolicy{Random._GLOBAL_RNG,
QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}}, NothingUpdater}, Random._GLOBAL_RNG}, Random._GLOBAL_RNG},
Random._GLOBAL_RNG, Int64}, is::Tuple{Int64, Int64})
    @ POMDPTools.Simulators
C:\Users\szhou\.julia\packages\POMDPTools\0tgtD\src\Simulators\stepthrough.jl:51
  [7] _collect(cont::UnitRange{Int64},
itr::POMDPTools.Simulators.MDPSimIterator{(:s, :a, :sp, :r, :info, :t,
:action_info), QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"),
Int64, Int64, NamedTuple{(:stateindex, :isterminal, :actionindex,
:initialstate, :gen, :actions, :states, :discount), Tuple{Dict{Int64,
Int64}, Bool, Dict{Int64, Int64}, var"#30#33", var"#29#32",
UnitRange{Int64}, UnitRange{Int64}, Float64}}},
MCTSPlanner{QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64,
Int64, NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate,
:gen, :actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool,
Dict{Int64, Int64}, var"#30#33", var"#29#32", UnitRange{Int64},
UnitRange{Int64}, Float64}}}, Int64, Int64,
MCTS.SolvedRolloutEstimator{RandomPolicy{Random._GLOBAL_RNG,
QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}}, NothingUpdater}, Random._GLOBAL_RNG}, Random._GLOBAL_RNG},
Random._GLOBAL_RNG, Int64}, #unused#::Base.HasEltype, isz::Base.SizeUnknown)
    @ Base .\array.jl:661
  [8] collect
    @ .\array.jl:649 [inlined]
  [9] collect_history(it::POMDPTools.Simulators.MDPSimIterator{(:s, :a,
:sp, :r, :info, :t, :action_info),
QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}},
MCTSPlanner{QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64,
Int64, NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate,
:gen, :actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool,
Dict{Int64, Int64}, var"#30#33", var"#29#32", UnitRange{Int64},
UnitRange{Int64}, Float64}}}, Int64, Int64,
MCTS.SolvedRolloutEstimator{RandomPolicy{Random._GLOBAL_RNG,
QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}}, NothingUpdater}, Random._GLOBAL_RNG}, Random._GLOBAL_RNG},
Random._GLOBAL_RNG, Int64}, cap_ex::Val{false}, prog::Nothing)
    @ POMDPTools.Simulators
C:\Users\szhou\.julia\packages\POMDPTools\0tgtD\src\Simulators\history_recorder.jl:185
 [10] simulate(sim::HistoryRecorder,
mdp::QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}},
policy::MCTSPlanner{QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"),
Int64, Int64, NamedTuple{(:stateindex, :isterminal, :actionindex,
:initialstate, :gen, :actions, :states, :discount), Tuple{Dict{Int64,
Int64}, Bool, Dict{Int64, Int64}, var"#30#33", var"#29#32",
UnitRange{Int64}, UnitRange{Int64}, Float64}}}, Int64, Int64,
MCTS.SolvedRolloutEstimator{RandomPolicy{Random._GLOBAL_RNG,
QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}}, NothingUpdater}, Random._GLOBAL_RNG}, Random._GLOBAL_RNG},
init_state::Int64)
    @ POMDPTools.Simulators
C:\Users\szhou\.julia\packages\POMDPTools\0tgtD\src\Simulators\history_recorder.jl:158
 [11] simulate(sim::HistoryRecorder,
mdp::QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}},
policy::MCTSPlanner{QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"),
Int64, Int64, NamedTuple{(:stateindex, :isterminal, :actionindex,
:initialstate, :gen, :actions, :states, :discount), Tuple{Dict{Int64,
Int64}, Bool, Dict{Int64, Int64}, var"#30#33", var"#29#32",
UnitRange{Int64}, UnitRange{Int64}, Float64}}}, Int64, Int64,
MCTS.SolvedRolloutEstimator{RandomPolicy{Random._GLOBAL_RNG,
QuickMDP{UUID("74f1b856-2074-4d8f-981b-34bec88d2e33"), Int64, Int64,
NamedTuple{(:stateindex, :isterminal, :actionindex, :initialstate, :gen,
:actions, :states, :discount), Tuple{Dict{Int64, Int64}, Bool, Dict{Int64,
Int64}, var"#30#33", var"#29#32", UnitRange{Int64}, UnitRange{Int64},
Float64}}}, NothingUpdater}, Random._GLOBAL_RNG}, Random._GLOBAL_RNG})
    @ POMDPTools.Simulators
C:\Users\szhou\.julia\packages\POMDPTools\0tgtD\src\Simulators\history_recorder.jl:137
 [12] top-level scope
    @ e:\Zhous\Dropbox\reinforcement learning for
maintenance\Matlab\MultipleUnits\multiunit1_gen_example.jl:103

**Comment by BoZenKhaa** [comment]
Ok, I think the issue is that you are using Integers for the state type. As a result, when[ `action_info` is called](https://github.com/JuliaPOMDP/MCTS.jl/blob/dd3531b32314628c68d849aeaad36872ab013141/src/vanilla.jl#L202-L206), instead of the correct [method that performs state id lookup](https://github.com/JuliaPOMDP/MCTS.jl/blob/dd3531b32314628c68d849aeaad36872ab013141/src/vanilla.jl#L133), the default [constructor for `StateNode` is used](https://github.com/JuliaPOMDP/MCTS.jl/blob/dd3531b32314628c68d849aeaad36872ab013141/src/vanilla.jl#L129-L132).

This seems to be a bug in MCTS. As a workaround, you can define a custom struct for your states, then the `StateNode` function will dispatch correctly.

**Comment by zhousy80** [comment]
Thanks for the note. I will follow the suggestion and try.

**Comment by BoZenKhaa** [comment]
You can also try the linked patch that adresses this issue.

**Comment by zhousy80** [comment]
Sorry about my ignorance, how can I get the linked patch? I added the MCTS
package again in my Julia and it does not make a difference.

**Comment by BoZenKhaa** [comment]
I guess the link is only visible on github. I mean [this](https://github.com/JuliaPOMDP/MCTS.jl/pull/100) pull request that is meant to address your issue.

**Comment by NeroBlackstone** [comment]
Same problem here. 
I'll check if the patched version works in my MDP.

**Comment by BoZenKhaa** [comment]
I will test the fix https://github.com/JuliaPOMDP/MCTS.jl/pull/100 again and merge later if it still works and no one objects. If you are using Integers for states that should fix it. You can also try wrapping your states in some other structs as a workaround.

**Comment by BoZenKhaa** [comment]
Closed by #100
